packages/core/src/modules/connections/services/ConnectionService.ts

```diff
@@ -18,286 +21,270 @@ import { AriesFrameworkError } from '../../../error'
+      .map((publicKey: PublicKey) => publicKey.value)
+      .filter((value): value is string => !!value)
+
+    const { did: peerDid } = await this.createDid({
+      role: DidDocumentRole.Received,
+      didDocument: convertToNewDidDocument(message.connection.didDoc),
+      recipientKeys: theirVerkeys.map(verkeyToDidKey),
+    })
 
-    connectionRecord.theirDidDoc = message.connection.didDoc
+    connectionRecord.theirDid = peerDid
     connectionRecord.theirLabel = message.label
     connectionRecord.threadId = message.id
-    connectionRecord.theirDid = message.connection.did
     connectionRecord.imageUrl = message.imageUrl
-
-    if (!connectionRecord.theirKey) {
-      throw new AriesFrameworkError(`Connection with id ${connectionRecord.id} has no recipient keys.`)
-    }
+    connectionRecord.outOfBandId = outOfBandRecord.id
 
     await this.updateState(connectionRecord, ConnectionState.Requested)
-
+    this.logger.debug(`Process message ${ConnectionRequestMessage.type} end`, connectionRecord)
     return connectionRecord
   }
 
   /**
    * Create a connection response message for the connection with the specified connection id.
    *
-   * @param connectionId the id of the connection for which to create a connection response
+   * @param connectionRecord the connection for which to create a connection response
    * @returns outbound message containing connection response
    */
   public async createResponse(
-    connectionId: string
+    connectionRecord: ConnectionRecord,
+    outOfBandRecord: OutOfBandRecord,
+    routing?: Routing
   ): Promise<ConnectionProtocolMsgReturnType<ConnectionResponseMessage>> {
-    const connectionRecord = await this.connectionRepository.getById(connectionId)
-
+    this.logger.debug(`Create message ${ConnectionResponseMessage.type} start`, connectionRecord)
     connectionRecord.assertState(ConnectionState.Requested)
     connectionRecord.assertRole(ConnectionRole.Inviter)
 
+    const { did } = routing ? routing : outOfBandRecord
+    if (!did) {
+      throw new AriesFrameworkError('Out-of-band record does not have did attribute.')
+    }
+
+    let didDocRouting = routing
+    if (!didDocRouting) {
+      didDocRouting = {
+        did,
+        endpoints: Array.from(
+          new Set(
+            outOfBandRecord.outOfBandMessage.services
+              .filter((s): s is DidCommService => typeof s !== 'string')
+              .map((s) => s.serviceEndpoint)
+          )
+        ),
+        verkey: didKeyToVerkey(outOfBandRecord.getTags().recipientKey),
+        routingKeys: Array.from(
+          new Set(
+            outOfBandRecord.outOfBandMessage.services
+              .filter((s): s is DidCommService => typeof s !== 'string')
+              .map((s) => s.routingKeys)
+              .filter((r): r is string[] => r !== undefined)
+              .reduce((acc, curr) => acc.concat(curr), [])
+              .map(didKeyToVerkey)
+          )
+        ),
```

Comment:
It's not nice. I would like to keep the support for more endpoints because I want to use a mediator with both HTTP and WebSockets.

What if I pass oob services if there are any and then do the following?

```
      const services = oobServices.map(
        (service, index) =>
          new IndyAgentService({
            id: `${did}#IndyAgentService`,
            serviceEndpoint: service.serviceEndpoint,
            recipientKeys: [verkey],
            routingKeys: service.routingKeys,
            priority: index,
          })
      )
```

Reply:
Hmm how did we previously do this? But yeah that approach sounds good to me. Maybe in the future we can update the routing object to already include services.

This will probably be refactored anyway when we add support for public dids, so yes your solutions sounds the easiest for now.